implement map, ap and chain for functions

benperez opened the pull request:
This change was motivated by some [discussion](https://github.com/plaid/sanctuary/issues/102) around porting some higher order Ramda functions into Sanctuary. The common theme among many of these functions is that they are essentially gluing together the results of different functions in different ways. If we implement Functor and Applicative for functions then we get this behavior for free.
```haskell
instance Functor ((->) r) where
    fmap = (.)

instance Applicative ((->) a) where
    pure = const
    (<*>) f g x = f x (g x)
```
The following simplifications become possible
 - `R.both` -> `R.lift(R.and)`
 - `R.either` -> `R.lift(R.or)`
 - `R.negate` -> `R.lift(R.not)`
 - `R.converge` -> `R.lift`

There are 2 issues that I ran into

What to do about converging functions that take multiple arguments, for example: 
```javascript
const f = R.converge((a, b) => a + b, a => a, (a, b) => a);
f.length; // 2
```
Right now `ap` just assumes there will only be one argument supplied to a function (but I could easily change it to apply the function to the entire arguments object). `R.converge` takes the max length of all of the functions but it can only do this because it knows the length of both input functions at the same time. If you were to perform the equivalent operation with `R.lift` you'd supply the input functions 1 at a time (although both at once is possible as well):
```javascript
const g = R.lift((a, b) => a + b); 
// g :: (Applicative f, Num c) => f c -> f c -> f c
const h = g(a => a); // should the function returned have length 1 now? How do you keep track of that? 
// h :: (Applicative f, Num c) => f c -> f c
const i = h((a, b) => a); // should the function returned have length 2 now?
```
I think the simplest solution is just to make `R.both`, `R.either`, etc. just take unary functions as arguments and call it a day.

The second problem is around preserving context but it seems like that has already been figured out for converge, I'm just not sure how to add it in to `R.map` and `R.ap`.

davidchambers commented:
OMG, this is some incredible @joneshf​-level jutsu!

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,25 @@ describe('liftN', function() {
+    var convergedOnBool = liftedAnd(R.gt(R.__, 0), R.lt(R.__, 3));
+    assert.strictEqual(typeof convergedOnInt, 'function');
+    assert.strictEqual(typeof convergedOnBool, 'function');
+    assert.deepEqual(convergedOnInt(10), (10 + 2) + (10 * 3) + (4 - 10));
```
We can use `assert.strictEqual` here (and below).

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,25 @@ describe('liftN', function() {
+
+  it('correctly interprets ((->) r) as a functor', function() {
+    var convergedOnInt = addN3(R.add(2), R.multiply(3), R.subtract(4));
+    var liftedAnd = R.liftN(2, R.and);
```
We could use `R.lift(R.and)` here, I believe.

davidchambers commented on src/internal/_map.js:
```diff
@@ -1,8 +1,14 @@
+  if (functor instanceof Function) {
+    return function(x) {
+      return fn(functor(x));
+    };
```
This is just `compose`! :smile:

davidchambers commented on src/ap.js:
```diff
@@ -19,8 +19,14 @@ var map = require('./map');
+           _reduce(function(acc, f) {
+             return _concat(acc, map(f, fn));
+           }, [], functor);
 });
```
Perhaps this would read better formatted like so:

```javascript
module.exports = _curry2(function ap(ftor, fn) {
  return (
    _hasMethod('ap', ftor) ?
      ftor.ap(fn) :
    ftor instanceof Function ?
      function(x) { return ftor(x)(fn(x)); } :
    // else
      _reduce(function(acc, f) { return _concat(acc, map(f, fn)); }, [], ftor)
  );
});
```

Also, let's use `typeof ftor === 'function'` rather than `ftor instanceof Function` for the reason given in #1140.

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,25 @@ describe('liftN', function() {
+  it('correctly interprets ((->) r) as a functor', function() {
+    var convergedOnInt = addN3(R.add(2), R.multiply(3), R.subtract(4));
+    var liftedAnd = R.lift(R.and);
+    var convergedOnBool = liftedAnd(R.gt(R.__, 0), R.lt(R.__, 3));
```
Let's remove the `liftedAnd` variable. `liftedAnd` doesn't convey any more information than `R.lift(R.and)`. ;)

CrossEye commented:
I love this.  I've always known that Haskell thinks of functions as functors over composition, although I didn't remember the applicative definition.

I hadn't ever _considered_ the possibility of including them here.  Nice done!  :fireworks: 

As to the two issues you raise,  I think we'll have to think pretty hard about both of them.  I'm quite loathe to go down the everything-is-unary route.  I don't know that we'd want to change the functions' APIs simply because we found an elegant new implementation that would demand it.

The `this` context has always been to me something very nice to have so long as it doesn't interfere too much with what else we want to do.  It shouldn't drive too much of our design.  But it _is_ nice to have, so as long as we can reasonably accommodate it, I feel we should.

But in any case, I would still want to extend `map` and `ap` this way; those are beautiful.

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,25 @@ describe('liftN', function() {
   });
+
+  it('correctly interprets [] as a functor', function() {
+    assert.deepEqual(addN3([1, 2, 3], [1, 2], [1, 2, 3]), [3, 4, 5, 4, 5, 6, 4, 5, 6, 5, 6, 7, 5, 6, 7, 6, 7, 8]);
```
It'd be easier to understand how the output is derived given `[1, 2, 3]`, `[10, 20]`, and `[100, 200, 300]` as inputs.

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,25 @@ describe('liftN', function() {
+    assert.deepEqual(addN3([1, 2], [3, 4], [5, 6]), [9, 10, 10, 11, 10, 11, 11, 12]);
+  });
+
+  it('correctly interprets ((->) r) as a functor', function() {
```
What does `((->) r)` indicate?

benperez replied:
It is the function type constructor with its first argument partially applied. Instances of Functor and Applicative must have type constructors that take exactly one argument, so we need to partially apply the first argument to `(->)`

davidchambers replied:
I'd like to state this in a way that makes sense in the context of JavaScript. Would it be accurate to state that `liftN` _interprets `a -> b` as a functor_? It seems to me we're trying to say that unary functions are functors.

benperez replied:
My understanding is that the return type of the function is the important part.
```javascript
var both = R.liftN(2, R.and);
// both :: Applicative f => f Bool -> f Bool -> f Bool
```
When we're talking about functions as _Applicative Functors_ we interpret 
`f Bool` to mean `((->) r) Bool` or `r -> Bool` 
so one interpretation of the type of the type of `both` is
```haskell
both :: (r -> Bool) -> (r -> Bool) -> (r -> Bool)
```
> Would it be accurate to state that liftN interprets a -> b as a functor? It seems to me we're trying to say that unary functions are functors.

In haskell I'm pretty sure that only unary functions can be applicative functors. In Javascript that's not the case, since we can hack around the arguments object to accept whatever number of arguments we want. We could indicate this by changing the description to something like:
> correctly interprets (* ->) as a functor

davidchambers replied:
I'll let you decide what makes the most sense. If that's sticking with `((->) r)`, that's fine with me.

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,24 @@ describe('liftN', function() {
     assert.deepEqual(addM(Maybe(3), Maybe(5)), Maybe(8));
   });
+
+  it('correctly interprets [] as a functor', function() {
```
The word "correctly" is redundant.

davidchambers commented on test/liftN.js:
```diff
@@ -50,4 +44,24 @@ describe('liftN', function() {
+
+  it('correctly interprets ((->) r) as a functor', function() {
+    var convergedOnInt = addN3(R.add(2), R.multiply(3), R.subtract(4));
+    var convergedOnBool = R.lift(R.and)(R.gt(R.__, 0), R.lt(R.__, 3));
```
I just realized that we're in __test/liftN.js__, so we should use `R.liftN` here after all. Sorry!

benperez commented:
@CrossEye I've updated this PR to do 2 things: 

1. addressed arity issues in `R.map` and `R.ap` for functions
2. switched the existing implementations of `R.complement`, `R.both`, and `R.either` to use `R.lift`.

Number 1 means that lifted functions can operate on functions of any length, which is consistent with the behavior of converged functions now and doesn't require Ramda to support "unary functions only".

Number 2 preserves the existing behavior of these functions with a cleaner, more generalized implementation __and__ has the additional benefit of generalizing them for all applicatives (in the case of `R.both` and `R.either`) and functors (in the case of `R.complement`).

I'm going to wait for something addressing https://github.com/ramda/ramda/issues/1401 to land before tackling the issue of context in lifted functions, especially with regards to `R.converge`.

davidchambers commented:
> switched the existing implementations of `R.complement`, `R.both`, and `R.either` to use `R.lift`

These definitions are absolutely beautiful! :heart:

CrossEye commented:
Fantastic!

Any objections, anyone?

kedashoe commented:
Amazing! Still trying to understand exactly what is going on, but a few comments:

- I believe the descriptions for `both` and `either` need updating as they do not short circuit anymore
- Should the descriptions for `map` and `ap` also be updated to mention this new behavior when they get functions?
- In `ap` we have the variable `ftor`, in `map` it is called `functor`
- Should we not include some tests in `test/map.js` and `test/ap.js`? Possibly (probably?) anything that breaks this behavior would be caught by the tests in `liftN`, but maybe appropriate to include tests closer to the source?

Great stuff!

scott-christopher commented:
This looks good to me.

Is wanting a `chain` implementation for functions going too far? :smile: 
```js
module.exports = _curry2(_dispatchable('chain', _xchain, function chain(fn, list) {
  if (typeof list == 'function') {
    return function(env) {
      return fn(list(env))(env);
    };
  }
  return _makeFlat(false)(map(fn, list));
}));
```

I would also be interested in trying to consolidate the dispatching, but don't really want to hijack this PR in the process.

davidchambers commented:
Great points, @kedashoe.

> Is wanting a `chain` implementation for functions going too far?

We can do that too? Cool! What do you say, @benperez?

benperez commented:
@kedashoe Thanks, I'll update the descriptions for `both` &  `either` &  `map` & `ap` and update the naming inconsistency. I'll also add tests in `map` and `ap`.

@scott-christopher Very nice! Will add that in tomorrow morning!

CrossEye commented:
> I believe the descriptions for both and either need updating as they do not short circuit anymore

That's actually a bit of a bummer.

It was one of the nice things about these functions.  But perhaps it wasn't really a particularly functional concern, just a standard JS one.

buzzdecafe commented on src/ap.js:
```diff
@@ -19,8 +21,15 @@ var map = require('./map');
+      ftor.ap(fn) :
+    typeof ftor === 'function' ?
+      curryN(Math.max(ftor.length, fn.length), function() {
+        return ftor.apply(this, _slice(arguments))(fn.apply(this, _slice(arguments)));
```
no need to slice twice

davidchambers replied:
No need to slice *at all*. :)

buzzdecafe replied:
indeed, even better

paldepind commented:
I think this is a good idea. But if we take this step we should go further still.

Objects for instance are functors.

Many other instances exists for native JavaScript data structures. I implemented a good bunch of them a while back in [seamless-fantasy](https://github.com/paldepind/seamless-fantasy).

davidchambers commented:
> I think this is a good idea. But if we take this step we should go further still.
>
> Objects for instance are functors.

I agree. Could you open an issue for discussion?

benperez commented:
I've updated this PR to address all comments.

buzzdecafe commented:
:cow2:

davidchambers commented on src/complement.js:
```diff
@@ -11,6 +11,8 @@ var _curry1 = require('./internal/_curry1');
  *   - applying `g` to zero or more arguments will give __false__ if applying
  *     the same arguments to `f` gives a logical __true__ value.
  *
+ * `R.not` will work on all other functors as well.
```
Should this be in the docstring for `R.not` instead?

benperez replied:
Good catch, should have said `R.complement`. :zap:

CrossEye commented:
>> Objects for instance are functors.

> I agree. Could you open an issue for discussion?

Yes, definitely worth discussing, but also should be a separate issue.

davidchambers commented:
:deciduous_tree:

davidchambers commented:
> Objects for instance are functors.

#1426

mattS-H commented:
The loss of short-circuiting is a real issue. I don't know why this has been ignored. For me its a performance problem primarily.

svozza commented:
https://github.com/ramda/ramda/pull/1475
